src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -0,0 +1,287 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import itertools
+import logging
+import os
+from pathlib import Path
+from typing import Any, Iterable, Mapping, Optional
+
+import toml
+from packaging.version import InvalidVersion, Version
+from pkg_resources import Requirement
+
+from pants.base.build_environment import get_buildroot
+
+logger = logging.getLogger(__name__)
+
+
+def get_max_caret(proj_name: str, version: str) -> str:
+    major = "0"
+    minor = "0"
+    micro = "0"
+
+    try:
+        parsed_version = Version(version)
+    except InvalidVersion:
+        logger.warning(
+            f"Warning: version {version} for {proj_name} is not PEP440-compliant; this requirement"
+            f" will be left as >={version},<{version}"
+        )
+        return version
+
+    if parsed_version.major != 0:
+        major = str(parsed_version.major + 1)
+    elif parsed_version.minor != 0:
+        minor = str(parsed_version.minor + 1)
+    elif parsed_version.micro != 0:
+        micro = str(parsed_version.micro + 1)
+    else:
+        base_len = len(parsed_version.base_version.split("."))
+        if base_len >= 3:
+            micro = "1"
+        elif base_len == 2:
+            minor = "1"
+        elif base_len == 1:
+            major = "1"
+
+    return f"{major}.{minor}.{micro}"
+
+
+def get_max_tilde(proj_name: str, version: str) -> str:
+    major = "0"
+    minor = "0"
+    micro = "0"
```

Comment:
No need to define `micro`, you can inline it in the f-string because it never gets changed.